amplipi/audiodetector/audiodetector.c

```diff
@@ -0,0 +1,354 @@
+  status->ch3r = (max_data->ch3r - baseline->ch3r) > CH_ACTIVE_THR;
+}
+
+// Given the window of past raw data, populate maxvals with the max along
+// channels
+void compute_window_max(AdcWindow *window, AdcData *maxvals) {
+  // iterate through channels
+  for (size_t i = 0; i < sizeof(maxvals->vals) / sizeof(maxvals->vals[0]);
+       i++) {
+    // iterate through samples
+    maxvals->vals[i] = 0;
+    for (size_t j = 0; j < sizeof(window->ch0l) / sizeof(window->ch0l[0]);
+         j++) {
+      uint16_t curr_val = window->vals[i * WINDOW_SIZE + j];
+      if (curr_val > maxvals->vals[i]) {
+        maxvals->vals[i] = curr_val;
+      }
+    }
+  }
+}
+
+void print_data(AdcData *data) {
+  printf("{ %u, %u, %u, %u, %u, %u, %u, %u }\n", data->ch0l, data->ch0r,
+         data->ch1l, data->ch1r, data->ch2l, data->ch2r, data->ch3l,
+         data->ch3r);
+  // fflush(stdout);
+}
+
+void print_status(AudioStatus *status) {
+  printf("{ %u, %u, %u, %u, %u, %u, %u, %u }\n", status->ch0l, status->ch0r,
+         status->ch1l, status->ch1r, status->ch2l, status->ch2r, status->ch3l,
+         status->ch3r);
+}
+
+void print_csv_header(FILE *csv) {
+  fprintf(csv,
+          "ch0l,ch0r,ch1l,ch1r,ch2l,ch2r,ch3l,ch3r,s0l,s0r,s1l,s1r,s2l,s2r,s3l,"
+          "s3r,\n");
+}
+
+void print_to_csv(AdcData *data, AudioStatus *status, FILE *csv) {
+#ifdef DUMP_CSV
+  fprintf(csv, "%u,%u,%u,%u,%u,%u,%u,%u,%u,%u,%u,%u,%u,%u,%u,%u,\n", data->ch0l,
+          data->ch0r, data->ch1l, data->ch1r, data->ch2l, data->ch2r,
+          data->ch3l, data->ch3r, status->ch0l, status->ch0r, status->ch1l,
+          status->ch1r, status->ch2l, status->ch2r, status->ch3l, status->ch3r);
+#endif
+}
+
+FILE *open_csv() {
+#ifdef DUMP_CSV
+  FILE *f_csv;
+  f_csv = fopen("raw_dump.csv", "w+");
+  print_csv_header(f_csv);
+  return f_csv;
+#else
+  return NULL;
+#endif
+}
+
+void close_csv(FILE *f_csv) {
+#ifdef DUMP_CSV
+  fclose(f_csv);
+#endif
+}
+
+void print_values(AdcData *baseline, AdcData *raw_data, AdcData *max_data,
+                  AudioStatus *status) {
+#ifdef DEBUG_PRINT
+  printf("baseline:    ");
+  print_data(baseline);
+  printf("raw data:    ");
+  print_data(raw_data);
+  printf("window max:  ");
+  print_data(max_data);
+  printf("status:      ");
+  print_status(status);
+#endif
+}
+
+int main(int argc, char **argv) {
+  bool success = register_sig_handler();
+  if (!success) {
+    fprintf(stderr, "Failed to register SIGINT handler\n");
+    return EXIT_FAILURE;
+  }
+
+  int status_fd;
+  success = open_storage(STATUS_FILE_PATH, &status_fd);
+  if (!success) {
+    fprintf(stderr, "Failed to create/open status file\n");
+    return EXIT_FAILURE;
+  }
+
+  AdcData baseline;  // Minimum values ever recorded (of window max)
+  for (size_t i = 0; i < sizeof(AdcData) / sizeof(baseline.vals[0]); i++)
+    baseline.vals[i] = DEFAULT_BASELINE;
+  AdcWindow window;
+  memset(&window, 0xFF, sizeof(AdcWindow));
```

Comment:
max value is safest for the initial state of the window. we check if the max value across the window is above/below baseline + some offset. I guess memset was used instead of a for loop because the whole thing can just be initialized to 0xFF.